File: workspace/planilha/parser.js
const pdfParse = require('pdf-parse');
const fs = require('fs');

// Converte "1.234,56" ou "-1.234,56" para número
function parseValor(str) {
  return parseFloat(str.replace(/\./g, '').replace(',', '.'));
}

// Formata data DD/MM/YYYY → YYYY-MM-DD (para ordenar no Sheets)
function formatData(str) {
  const [d, m, y] = str.split('/');
  return `${y}-${m}-${d}`;
}

// Extrai mês como YYYY-MM para agrupar no dashboard
function extrairMes(dataISO) {
  return dataISO.slice(0, 7);
}

async function parsePDF(caminhoArquivo) {
  const buffer = fs.readFileSync(caminhoArquivo);
  const result = await pdfParse(buffer);
  return result.text;
}

// ── Parser Itaú Conta Corrente ───────────────────────────────────────────────
// Formato esperado no texto extraído:
//   DD/MM/YYYY DESCRIÇÃO VALOR DD/MM/YYYY ... (tudo concatenado)
// Estratégia: split por marcadores de data, depois extrai valor do final
function parseItauCC(texto) {
  // Localiza o início dos lançamentos (depois do header da tabela)
  const marcador = 'saldo (R$)';
  const idx = texto.toLowerCase().indexOf(marcador);
  const conteudo = idx >= 0 ? texto.slice(idx + marcador.length) : texto;

  const transacoes = [];

  // Regex: captura DATE + DESCRIPTION + AMOUNT
  // Usa lookahead para parar na próxima data ou no fim
  const rx = /(\d{2}\/\d{2}\/\d{4})\s+([\s\S]+?)\s+([-]?\d{1,3}(?:\.\d{3})*,\d{2})(?=\s+\d{2}\/\d{2}\/\d{4}|\s*$)/g;

  let match;
  while ((match = rx.exec(conteudo)) !== null) {
    const [, data, descRaw, valorStr] = match;
    const descricao = descRaw.replace(/\s+/g, ' ').trim();

    // Ignora linhas de saldo e rendimentos mínimos irrelevantes
    if (/SALDO DO DIA/i.test(descricao)) continue;
    if (/REND PAGO APLIC AUT MAIS/i.test(descricao) && Math.abs(parseValor(valorStr)) < 10) continue;

    const valor = parseValor(valorStr);
    const dataISO = formatData(data);

    transacoes.push({
      data: dataISO,
      dataOriginal: data,
      descricao,
      valor,
      mes: extrairMes(dataISO),
    });
  }

  return transacoes;
}

// ── Mapa de parsers por tipo de conta ────────────────────────────────────────
const PARSERS = {
  'Itaú CC': parseItauCC,
  // Adicionar quando chegar os PDFs:
  // 'Itaú Cartão': parseItauCartao,
  // 'BTG Cartão':  parseBTGCartao,
};

async function processarExtrato(caminhoPDF, tipoConta) {
  const parser = PARSERS[tipoConta];
  if (!parser) {
    throw new Error(`Tipo de conta desconhecido: "${tipoConta}". Disponíveis: ${Object.keys(PARSERS).join(', ')}`);
  }
  const texto = await parsePDF(caminhoPDF);
  return parser(texto);
}

module.exports = { processarExtrato, parseValor, formatData };

File: workspace/planilha/parser.py
#!/usr/bin/env python3
"""
parser.py — Extrai transações de extratos bancários em PDF
Uso: python3 parser.py <caminho_pdf> <tipo_conta>
Tipos: "Itaú CC" | "Itaú Cartão" | "BTG Cartão"

Retorna JSON array de transações para o stdout.
"""

import sys
import json
import re
import pdfplumber


def parse_valor(s: str) -> float:
    """Converte '1.234,56' ou '-1.234,56' para float."""
    return float(s.replace('.', '').replace(',', '.'))


def format_data(s: str) -> str:
    """Converte 'DD/MM/YYYY' para 'YYYY-MM-DD'."""
    d, m, y = s.split('/')
    return f"{y}-{m}-{d}"


RE_DATA = re.compile(r'^\d{2}/\d{2}/\d{4}$')
RE_VALOR = re.compile(r'^-?\d{1,3}(?:\.\d{3})*,\d{2}$')

# Descrições a ignorar (não são transações)
IGNORAR = re.compile(
    r'(SALDO DO DIA|REND PAGO APLIC AUT MAIS|^data$|^lançamentos$|saldo \(R\$\)|valor \(R\$\))',
    re.IGNORECASE
)


def extrair_linhas(caminho_pdf: str) -> list[dict]:
    """
    Usa pdfplumber com word-level extraction para reconstruir linhas.
    Retorna lista de {'data', 'descricao', 'valor'}.
    """
    transacoes = []

    with pdfplumber.open(caminho_pdf) as pdf:
        for page in pdf.pages:
            words = page.extract_words(x_tolerance=4, y_tolerance=3)

            # Agrupa palavras por posição y (±5px = mesma linha)
            linhas: dict[int, list] = {}
            for w in words:
                y_key = round(w['top'] / 6) * 6
                linhas.setdefault(y_key, []).append(w)

            for y in sorted(linhas.keys()):
                ws = sorted(linhas[y], key=lambda w: w['x0'])
                tokens = [w['text'] for w in ws]

                if not tokens:
                    continue

                # Linha começa com data?
                if not RE_DATA.match(tokens[0]):
                    continue

                data_str = tokens[0]
                resto = tokens[1:]

                if not resto:
                    continue

                # Último token é um valor?
                ultimo = resto[-1]
                if not RE_VALOR.match(ultimo):
                    continue

                descricao = ' '.join(resto[:-1])
                valor_str = ultimo

                # Ignora linhas que não são transações reais
                if IGNORAR.search(descricao):
                    continue

                # Ignora rendimentos automáticos < R$10 (irrelevante)
                valor = parse_valor(valor_str)
                if abs(valor) < 1.0:
                    continue

                transacoes.append({
                    'data': format_data(data_str),
                    'dataOriginal': data_str,
                    'descricao': descricao,
                    'valor': valor,
                    'mes': format_data(data_str)[:7],
                })

    return transacoes


def parse_itau_cc(caminho_pdf: str) -> list[dict]:
    return extrair_linhas(caminho_pdf)


PARSERS = {
    'Itaú CC': parse_itau_cc,
    # Adicionar quando chegar os PDFs:
    # 'Itaú Cartão': parse_itau_cartao,
    # 'BTG Cartão':  parse_btg_cartao,
}


def main():
    if len(sys.argv) < 3:
        print('Uso: python3 parser.py <caminho_pdf> "<tipo_conta>"', file=sys.stderr)
        sys.exit(1)

    caminho_pdf = sys.argv[1]
    tipo_conta = sys.argv[2]

    if tipo_conta not in PARSERS:
        print(f'Tipo de conta desconhecido: "{tipo_conta}". Disponíveis: {list(PARSERS.keys())}', file=sys.stderr)
        sys.exit(1)

    transacoes = PARSERS[tipo_conta](caminho_pdf)
    print(json.dumps(transacoes, ensure_ascii=False, indent=2))


if __name__ == '__main__':
    main()
